akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -74,11 +76,38 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
 
   private def updateState(streamId: Int, handle: IncomingStreamState => IncomingStreamState): Unit = {
     val oldState = streamFor(streamId)
-    val newState = handle(oldState)
+    val newState: IncomingStreamState = handle(oldState)
+
+    // only validate maxConcurrentStreams compliance if we'd increment the number
```

Comment:
I'd like to keep this method simple. Could we find a better place to put this logic? E.g. we know that only HEADERS and PUSH_PROMISE frames can create a stream, so we can put the logic into `handleStreamEvent`.

Reply:
And then let's move the enforcing logic into `Http2ServerDemux` and just call that with the current number of open streams.